docs/rfc/rfc-003-performance-questions.md

```diff
@@ -0,0 +1,193 @@
+that are often cited as having performance issues. It raises questions and suggests
+lines of inquiry that may be valuable for better understanding Tendermint's performance issues.
+
+As a note: We should avoid quickly adding many microbenchmarks or package level benchmarks. 
+These are prone to being worse than useless as they can obscure what _should_ be  
+focused on: performance of the system from the perspective of a user. We should, 
+instead, tune performance with an eye towards actions user needs.
+
+REQUEST: The list below may be incomplete, if there are additional sections that are often
+cited as creating poor performance, please comment so that they may be included.
+
+### P2P
+
+#### Tendermint cannot scale to large numbers of nodes
+
+A complaint has been reported that Tendermint networks cannot scale to large numbers of nodes.
+The listed number of nodes a user reported as causing issue was in the thousands.
+We don't currently have evidence about what the upper-limit of nodes that Tendermint's
+P2P stack can scale to.
+
+We need to more concretely understand the source of issue and determine what layer
+is causing a problem. It's possible that the P2P layer, in the absence of any reactors
+sending data, is perfectly capable of managing thousands of peer connections. For
+a reasonable networking and application setup, thousands of connections should not present any
+issue for the application.
+
+We need more data to understand the problem directly. We want to drive the popularity
+and adoption of Tendermint and this will mean allowing for chains with more validators.
+We should follow up with users experiencing this issue. We may then want to benchmark
+the P2P layer in a few simple ways:
+1. By starting a network with 1000 nodes and sending no messages to determine memory
+and CPU consumption in the presence of 1000 nodes.
+2. By testing two nodes connected to each other via P2P, and attempting to
+generate as much load between the two nodes as possible. This will help inform
+how much load is incurred from message sends in an otherwise unencumbered P2P stack.
+
+### Sync
+
+#### Block Syncing is slow
+
+Bootstrapping a new node in a network to the height of the rest of the network
+takes longer than users would like. Block sync requires fetching all of the blocks from
+peers and placing them into the local disk for storage. A useful line of inquiry
+is understanding how quickly a perfectly tuned system _could_ fetch all of the state
+over a network so that we understand how much overhead Tendermint actually adds.
+
+The operation is likely to be _incredibly_ dependent on the environment in which
+the node is being run. The factors that will influence syncing include:
+1. Number of peers that a syncing node may fetch from.
+2. Speed of the disk that a validator is writing to.
+3. Speed of the network connection between the different peers that node is
+syncing from.
```

Comment:
While yes, bandwidth is an overhead, the general design of fast sync should be changed. This overly focuses on bandwidth / speed of download. Whereas improving this category includes:
* Reducing computation time of Tendermint (improved header verification)
* Lowering bandwidth needed, by compressing block data / headers
* Lowering header size
* Parallelizing block verification

In addition to whats being suggested of "Ensuring we're downloading optimally".

Reply:
parallelization of block verification ++

Reply:
Thanks for the comment! I can definitely work to improve the language here to clarify the intent of this section. It does sounds like a number of the problems syncing faces may be protocol-level but in my view it would still be helpful to understand the current performance bottlenecks so we can accurately gauge improvements.

To clarify, I'm not exactly trying to ensure we're downloading optimally, i.e. shuffling bytes from the network rxqueues into process memory etc., but more looking for yardsticks we can use to gauge performance. Placing an upper bound may be illuminating in understanding how well we could perform a sync operation given the hard constraints of the system. Bandwidth and drive speeds, on large data sets, are likely to form the upper bounds. Knowing those limits will help us understanding how much _extra_ work tendermint is doing is when performing a sync.

Reply:
The current network we are operating. the nodes didn't hit the bandwidth limitation (far behind) of the cloud vendor is providing. But the sync process is still not fast. I will prefer some p2p multiplex / PEX logic that needs to be investigating.